src/composables/useExcelDownload.ts

```diff
@@ -1,138 +1,69 @@
-                        } else {
-                            row.push(mean, featureValues[indicatorId].sd);
-                        }
-                    });
-                    sheetData.push(row);
-                }
-            });
-        });
-
-        const sheet = XLSX.utils.aoa_to_sheet(sheetData);
-        XLSX.utils.book_append_sheet(workbook, sheet, `admin${level}`);
-    };
-
-    const buildGlobalIndicatorsWorkbook = (workbook: WorkBook) => {
-        writeTab(workbook, 1, admin1Indicators.value, admin1Geojson.value);
-    };
-
-    const buildCountryIndicatorsWorkbook = (workbook: WorkBook, countryId: string) => {
-        writeTab(workbook, 1, admin1Indicators.value, admin1Geojson.value, countryId);
-
-        const { countriesWithoutAdmin2 } = appConfig.value;
-        const admin1Only = countriesWithoutAdmin2.includes(countryId);
-        if (!admin1Only) {
-            writeTab(workbook, 2, admin2Indicators.value, admin2Geojson.value, countryId);
-        }
-    };
-
-    const downloadFile = () => {
+    const doDownloadSelectedCountry = () => {
         const workbook = XLSX.utils.book_new();
         const { country } = mapSettings.value;
-        const fileName = `arbomap_${country || "GLOBAL"}.xlsx`;
-        if (country) {
-            buildCountryIndicatorsWorkbook(workbook, country);
-        } else {
-            buildGlobalIndicatorsWorkbook(workbook);
+        if (!country) {
+            throw Error("No selected country");
         }
+        const fileName = excelFilename(country);
+        const builder = new BuildExcel(
+            appConfig.value,
+            countryNames.value,
+            admin1Indicators.value,
+            admin2Indicators.value,
+            admin1Geojson.value,
+            admin2Geojson.value
+        );
 
+        builder.buildCountryIndicatorsWorkbook(workbook, country);
         XLSX.writeFile(workbook, fileName);
     };
 
-    const download = () => {
+    const doDownloadGlobal = async (includeAdmin2: boolean) => {
+        const filename = excelFilename(null, includeAdmin2 ? AdminLevel.TWO : AdminLevel.ONE);
+        const url = `${APP_BASE_URL}/resources/excel/${filename}`;
+        await downloadFile(url, filename);
+    };
+
+    const download = (doDownload: () => void | Promise<void>) => {
         downloadError.value = null;
-        debounce(() => {
+        debounce(async () => {
             try {
-                downloadFile();
+                await doDownload();
             } catch (e) {
                 console.log(`Error downloading Excel file: ${e}`);
                 downloadError.value = e;
             }
         })();
     };
 
-    return { download, downloadError };
+    const downloadSelectedCountry = () => {
+        download(doDownloadSelectedCountry);
+    };
+
+    const downloadGlobal = (includeAdmin2: bool) => {
+        download(async () => {
+            await doDownloadGlobal(includeAdmin2);
+        });
```

Comment:
Having all these similarly named functions (download, downloadGlobal, doDownloadGlobal, doDownload) gets confusing to read. I would suggest, instead of having a separate function called doDownloadGlobal, we put the implementation in downloadGlobal:

```suggestion
        download(async () => {
            const filename = excelFilename(null, includeAdmin2 ? AdminLevel.TWO : AdminLevel.ONE);
            const url = `${APP_BASE_URL}/resources/excel/${filename}`;
            await downloadFile(url, filename);
        });
```

And the equivalent for non-global (selected country).

What do you think?

Reply:
Yeah, will do. It does mean an extra level of nesting for the eyes to negotiate and brain to perceive, but I think you're right that it's probably better than a forest of download.. methods..